internal/graph/preloadedtaskrunner.go

```diff
@@ -102,3 +118,31 @@ func (tr *preloadedTaskRunner) selectTask() TaskFunc {
 	task, tr.tasks = tr.tasks[0], tr.tasks[1:]
 	return task
 }
+
+func (tr *preloadedTaskRunner) storeErrorAndCancel(err error) {
+	tr.lock.Lock()
+	defer tr.lock.Unlock()
+
+	if tr.err == nil {
+		tr.err = err
+		tr.cancel()
+	}
+}
+
+func (tr *preloadedTaskRunner) emptyForCancel() {
+	tr.lock.Lock()
+	defer tr.lock.Unlock()
+
+	if tr.err == nil {
+		tr.err = tr.ctx.Err()
+	}
+
+	for {
+		if len(tr.tasks) == 0 {
+			break
+		}
+
+		tr.tasks = tr.tasks[1:]
+		tr.wg.Done()
+	}
```

Comment:
I didn't think reslicing like this released the underlying array values until you re-assign over the same index?

maybe something like this?

```suggestion
    for i := 0; i < len(tr.tasks); i++ {
        tr.wg.Done()
    }
    tr.tasks = make([]TaskFunc, 0, len(tr.tasks))
```

Reply:
but I don't think it will release the functions that are pointed to by the underlying array?

seems like a small change in exchange for free more garbage faster?